Connector Configuration Overview and Studio Topics (#546)
mulesoft/docs-connectors: modules/ROOT/pages/introduction/intro-config-use-studio.adoc

= Use Studio to Configure a Connector
ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> :imagesdir: ../../assets/images/

Reviewer: Remove this metadata. The _imagesdir_ attribute is already being set in the __attributes.adoc_ file.

To use Anypoint Studio to configure a connector:

. Install the connector.

> . Configure an input source for the connector.

Reviewer: I would say an "input source for the flow of the connector" or something like that. Connectors don't have input sources, flows do. Some connectors even have operations that can act as input sources for the flow.

. Configure the connector.

Follow these best practices when you configure an input source and connector in Studio:

*  Configure an associated global element for the input source and connector.
* Create a `.yaml` file or `.properties` file to contain your properties, and then reference the file from the configuration.
* Configure a reconnection strategy for the input source and target connector.

For information about these best practices, see <<introduction/intro-connector-configuration-overview.adoc#best-practices,Best Practices for Configuring Connectors>>.

== Install the Connector

> Installing the connector adds it to a Studio project. Studio provides two ways to install a connector:

Reviewer: There's one more, and actually the one that I would recommend users to use: The **Manage Mule Project** option: 
![manage-mule-project](https://user-images.githubusercontent.com/16763154/72365808-b6257c80-36d7-11ea-9451-d384720ea590.png)

This option lets users configure all the dependencies of their project:

![modules-management](https://user-images.githubusercontent.com/16763154/72366305-c2f6a000-36d8-11ea-9869-f1be35572191.png)

In the **Modules** option users can:
1. Add or remove modules/connectors.
2. Look in Exchange for newer versions of the connectors and update them from there.

Author: Leaving out for now because of the issue with the connector not showing up in the palette.

* Use the Exchange application in the Studio taskbar.

* Use the Exchange icon in the Mule Palette view.

> (*Reviewers, is there any advantage to using one of these methods for installing a connector* )

Reviewer 2: I can see only visual differences between the two.
Also it is possible to install the connector by manually adding a dependency into the `pom.xml`, in fact that's what the other two options do in the background.

Author: I am going to cover the pom.xml piece in a separate topic.

Author: I will be covering the pom.xml in a separate topic.

=== Install the Connector from the Studio Taskbar

. In Studio, create a Mule project.
. Click the Exchange icon *(X)* in the upper left of the Studio task bar.

> . In Exchange, click *Login* and supply your Anypoint Platform username and password.

Reviewer: The credentials settings should be outside these steps. When users add a new connector from the Mule Pallete view or Exchange, in both cases they'll need to provide credentials. 

I would move the Anypoint Platform configuration step as a prerequesite under **To use Anypoint Studio to configure a connector:**
You can link to this task in the Studio docs: https://docs.mulesoft.com/studio/7.4/set-credentials-in-studio-to

Also, this topic would only work for US cloud users. EU Cloud, and PCE users need to specifically configure Studio to point to their organizations: https://docs.mulesoft.com/studio/7.4/pce-configuration

. In Exchange, search for the connector, for example, *Amazon SQS*.
. Select the connector.
. Click *Add to project*.
. Follow the prompts to install the connector.

=== Install the Connector from the Mule Palette View

. In Studio, create a Mule project.

> . In the Mule Palette view, click *(X) Search in Exchange*.

Reviewer 3: Do we need "view"? Can we just say Mule Palette and make it bold, since it's an interface element?

Author: In past edits, Kim has wanted us to use "view."

Reviewer: I agree with that. The Mule palette is one of many Views in Studio.

. In *Add Modules to Project*, enter the connector name in the search field.
. Click the connector name in *Available modules*.
. Click *Add*.
. Click *Finish*.

> == Configure an Input Source for the Connector

Reviewer: Same as my comment above. You are configuring an input source for the flow, not the connector.

An input source starts a flow when a specified condition is met. Configure an input source for the connector, such as an HTTP Listener operation (HTTP Listener), Scheduler endpoint (Scheduler), or connector-specific input source.

=== Use HTTP Listener as an Input Source

HTTP Listener listens for HTTP requests and triggers the associated flow when it receives them. To use HTTP Listener as an input source:

. In the Mule Palette view, search for *HTTP* and select the *Listener* operation:
+

> image::introduction/intro-config-studio-select-listener.png[Select HTTP Listener as an Input Source]

Reviewer 3: Probably need a caption per screenshot guidelines.

Reviewer 3: (global)

Author: I'll be circling back on the graphics.

+
. Drag the *Listener* operation onto the canvas.
. In the *Listener* configuration screen, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for HTTP Listener.
+
The following screenshot shows the default values for the HTTP Listener global element:
+
image::introduction/intro-config-studio-listener.png[Global Element Screen for HTTP Listener]
+
. Accept the defaults on the *General* tab or specify required information:
* To listen on a specific port for every network interface on the server, leave the default value `0.0.0.0` for the *Host* property.
* To listen for incoming requests generated inside the server, set the *Port* value to `localhost`.

> . To use TLS, access the *TLS* tab and specify the required information.

Reviewer 3: ```suggestion
. To use TLS, click the *TLS* tab and specify the required information.
```

. On the *Advanced* tab, specify a reconnection strategy.
. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.

> . Set the *Path* field to the appropriate path.

Reviewer 3: such as?

. Select the other tabs and specify the appropriate field values.

For more information about HTTP Listener, see the xref:http/http-documentation.adoc[HTTP Connector Reference].

=== Use Scheduler as an Input Source

> Use Scheduler to trigger a flow when a time-based condition is met. You can configure Scheduler to trigger a flow at a regular interval or you can use a more flexible cron expression. For example, you can use Scheduler to trigger a flow every five seconds, or you can set up a cron expression to trigger a flow at noon everyday for one month.

Reviewer 3: Should this be "the Scheduler source" or something? Without an article, it's like a product name, but it's not a product name.

To use Scheduler as an input source:

. In the Mule Palette view, search for and select *Scheduler*:
+
image::introduction/intro-config-studio-select-scheduler.png[Select Scheduler as an Input Source]
+
. Drag *Scheduler* onto the canvas.
+
The following screenshot shows the default values for the scheduling strategy:
+
image::introduction/intro-config-studio-scheduler.png[Default Values for the Scheduling Strategy]
+

> . In the *Scheduling Strategy* field of the *Scheduler configuration* screen, select whether you want to use a fixed frequency or cron expression to trigger the flow

Reviewer 3: ```suggestion
. In the *Scheduling Strategy* field of the *Scheduler configuration* screen, select whether you want to use a fixed frequency or cron expression to trigger the flow.
```

* If you select *Fixed Frequency*, enter values for these fields:
+
** *Frequency*
+

> The frequency at which the Scheduler triggers the flow, based on the time unit specified in the *Time unit* field.

Reviewer 3: ```suggestion
The frequency at which the Scheduler triggers the flow, based on the time unit specified in the *Time unit* field
```

+
** *Start delay*
+

> The amount of time to wait before triggering the flow for the first time after the app starts, based on the time unit specified in the *Time unit* field.

Reviewer 3: ```suggestion
The amount of time to wait before triggering the flow for the first time after the app starts, based on the time unit specified in the *Time unit* field
```

Reviewer 3: No terminating punctuation for incomplete sentences.

+
** *Time unit*
+
The time unit for the values of *Frequency* and *Start Delay*.
+
* If you select *Cron*, enter the cron expression in the *Expression* field and the time zone in the *Time Zone* field. For

>  the format and valid values for the *Time Zone* field, see the Javadoc for the Java TimeZone class.

Reviewer 3: link?

Author: A good suggestion, but I am seeing lots of places this is documented, and don't know which specific one to refer to.

+
. To configure metadata for the payload, variables, and attributes, select the *Metadata* tab and add the appropriate metadata.

> For more information about Scheduler, see xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint (Trigger)]

Reviewer 3: ```suggestion
For more information about Scheduler, see xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint (Trigger)].
```

For example, you can use Amazon SQS Receive Messages as an input source for Amazon SQS Connector. This input source triggers a flow each time the configured Amazon SQS queue receives a new message.

To configure a connector-specific input source:

. In the Mule Palette view, search for the connector and select the input source.
+
For example, search for Amazon SQS and select the Receive Messages input source:
+
image::introduction/intro-config-studio-select-specific.png[Select Amazon SQS Receive Messages as an Input Source]
+
. Drag the input source onto the canvas.
. In the configuration screen, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for the input source.
. Enter values for the required fields.
. On the *Advanced* tab next to the *Proxy* tab, specify a reconnection strategy.
. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.

== Configure the Connector

To configure a connector in your Mule app, first configure a global element that can be used by all instances of that connector in the flow:

. In the Mule Palette view, search for the connector and select the desired operation.

> . Drag the operation onto the canvas to the right of the input source.2

Reviewer 3: ```suggestion
. Drag the operation onto the canvas to the right of the input source.
```

. In the configuration screen for the operation, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for connector.
. Specify the global element properties for the connector.
. On the *Advanced* tab next to the *Proxy* tab, specify a reconnection strategy.
. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.

After you configure a global element, configure the other fields for the connector.

> For information about configuring a specific connector, see the User Guide and Reference Guide for that connector.

Reviewer 3: lc since they're not actual titles?

> = Use Studio to Configure a Connector

Reviewer 4: Looking at the informal title for #546 PR, are we now in the "Studio Topics" rather than the "Configuration Overview"? Because what follows is task topics, not overview topics.

Author: Correct. This is showing a general path for using Studio to configure a connector.

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

To use Anypoint Studio to configure a connector:

. Supply your authentication credentials to Anypoint Platform.

> . Install the connector.

Reviewer 5: You aren't installing the connector, you are adding it to a project.

. Configure an input source for the connector's flow.
. Configure the connector.

Follow these best practices when you configure an input source and connector in Studio:

*  Configure an associated global element for the input source and connector.

> * Create a YAML file (`.yaml`) or properties file (`.properties`) to contain your properties, and then reference the file from the configuration.

Reviewer 4: ditto comment made in the Overview topic about this same sentence.

Reviewer 4: Are we omitting the connector test practice here? This list does not exactly match the similar one in Overview.

Author: Good point. I will add it here.

* Configure a reconnection strategy for the input source and target connector.

For information about these best practices, see <<introduction/intro-connector-configuration-overview.adoc#best-practices,Best Practices for Configuring Connectors>>.

== Supply Authentication Credentials to Anypoint Platform

> To access connectors on the Exchange portal, you must first configure Anypoint Platform credentials if these have not been configured:

Reviewer 4: ```suggestion
To access connectors on the Anypoint Exchange portal, you must first configure Anypoint Platform credentials if these have not been configured:
```

* If you use the US cloud, see xref:studio::set-credentials-in-studio-to.adoc[Configuring Anypoint Platform Credentials (Anypoint Studio)] for information.

> * If you use the EU cloud or Private Cloud Edition (PCE), see xref:studio::pce-configuration.adoc[Configure Private Cloud Edition in Studio] for information.

Reviewer 4: ```suggestion
* If you use the EU cloud or Anypoint Platform Private Cloud Edition (Anypoint PCE), see xref:studio::pce-configuration.adoc[Configure Private Cloud Edition in Studio] for information.
```

[[install]]

> == Install the Connector

Reviewer 5: I wouldn't call this installing a connector. You are just adding the connector to the project. In our template, we use "Add" not "Install".

Reviewer 4: ^

Installing a connector adds it to a Studio project:

> 

Reviewer 5: It's not really a "Studio project," which can be other types of projects. It's a Mule project.

Reviewer 4: Per Valkyrie's previous comment, if this task really should be "Add" rather than "Install," then reconsider this wording, including referring to a Mule project.

. In Studio, create a Mule project.
. In the Mule Palette view, click *(X) Search in Exchange*.
. In *Add Dependencies to Project*, enter the connector name in the search field.
. Click the connector name in *Available modules*.
. Click *Add*.
. Click *Finish*.

[[configureInputSource]]
== Configure an Input Source for the Connector

An input source initiates a flow when a specified condition is met. Configure an input source for the connector's flow, such as an HTTP Listener operation (HTTP Listener), Scheduler endpoint (Scheduler), or connector-specific input source.

> === Use HTTP Listener as an Input Source

Reviewer 5: Why are we putting so many procedures in a concept topic? I thought this was an overview. These detailed procedures are already in the doc for the specific connectors.

Author: This topic is not the overview topic. It is called Use Studio to Configure a Connector.

HTTP Listener listens for HTTP requests and triggers the associated flow when it receives them. To use HTTP Listener as an input source:

. In the Mule Palette view, search for *HTTP* and select the *Listener* operation:
+
.HTTP Listener initiates the associate flow when it receives HTTP requests

.Select HTTP Listener to listen for HTTP requests
image::introduction/intro-config-studio-select-listener.png[Selecting HTTP Listener as an input source]
+
. Drag the *Listener* operation onto the canvas.
. In the *Listener* configuration screen, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for HTTP Listener.
+

> The following screenshot shows the default values on the General tab for the HTTP Listener global element:

Reviewer 4: ```suggestion
The following screenshot shows the default values on the *General* tab for the HTTP Listener global element:
```

+
.Default values for the HTTP Listener global element
image::introduction/intro-config-studio-listener.png[HTTP Listener global element default values]
+
* The first section shows the default values for the protocol, host, and port to use in the connection.

> * The section section shows the default values for the base path and listener interceptors to use in the connection.

Reviewer 4: ```suggestion
* The second section shows the default values for the base path and listener interceptors to use in the connection.
```

+
. Accept the defaults on the *General* tab or specify required information:
* To listen on a specific port for every network interface on the server, leave the default value `0.0.0.0` for the *Host* property.
* To listen for incoming requests generated inside the server, set the *Port* value to `localhost`.
. To use TLS, click the *TLS* tab and specify the required information.
. On the *Advanced* tab, specify a reconnection strategy.
. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.
. Set the *Path* field to the appropriate path.
. Select the other tabs and specify the appropriate field values.

For more information about HTTP Listener, see the xref:http/http-documentation.adoc[HTTP Connector Reference].

=== Use Scheduler as an Input Source

> Use Scheduler to initiated a flow when a time-based condition is met. You can configure Scheduler to initiate a flow at a regular interval or you can use a more flexible cron expression. For example, you can use Scheduler to initiate a flow every five seconds, or you can set up a cron expression to initiate a flow at noon everyday for one month.

Reviewer 5: ```suggestion
Use Scheduler to initiate a flow when a time-based condition is met. You can configure Scheduler to initiate a flow at a regular interval or you can use a more flexible cron expression. For example, you can use Scheduler to initiate a flow every five seconds, or you can set up a cron expression to initiate a flow at noon every day for one month.
```

Reviewer 5: everyday is an adjective

> Use Scheduler to initiated a flow when a time-based condition is met. You can configure Scheduler to initiate a flow at a regular interval or you can use a more flexible cron expression. For example, you can use Scheduler to initiate a flow every five seconds, or you can set up a cron expression to initiate a flow at noon everyday for one month.

Reviewer 4: ```suggestion
Use Scheduler to initiate a flow when a time-based condition is met. You can configure Scheduler to initiate a flow at a regular interval or you can use a more flexible cron expression. For example, you can use Scheduler to initiate a flow every five seconds, or you can set up a cron expression to initiate a flow at noon every day for one month.
```

To use Scheduler as an input source:

> . In the Mule Palette view, search for and select *Scheduler*:

Reviewer 4: ```suggestion
. In the *Mule Palette* view, search for and select *Scheduler*:
```

+
.Select Scheduler to schedule when a flow is initiated
image::introduction/intro-config-studio-select-scheduler.png[Selecting Scheduler as an input source]
+
. Drag *Scheduler* onto the canvas.
+
The following screenshot shows the default values for the scheduling strategy:
+
.Default values for the scheduling strategy
image::introduction/intro-config-studio-scheduler.png[Default values for the scheduling strategy]
+

> . In the *Scheduling Strategy* field of the *Scheduler configuration* screen, select whether you want to use a fixed frequency or cron expression to initiate the flow.

Reviewer 4: ```suggestion
. In the *Scheduling Strategy* field of the *Scheduler configuration* screen, select whether you want to use a fixed frequency or cron expression to initiate the flow:
```

* If you select *Fixed Frequency*, enter values for these fields:
+
** *Frequency*
+
The frequency at which the Scheduler initiates the flow, based on the time unit specified in the *Time unit* field
+
** *Start delay*
+

> The amount of time to wait before initiating the flow for the first time after the app starts, based on the time unit specified in the *Time unit* field

Reviewer 5: ```suggestion
The amount of time to wait before initiating the flow for the first time after the app starts, based on the time unit specified in the *Time unit* field.
```

+
** *Time unit*
+
The time unit for the values of *Frequency* and *Start Delay*
+

> * If you select *Cron*, enter the cron expression in the *Expression* field and the time zone in the *Time Zone* field. For

Reviewer 4: GLOBAL re. list items and steps:
```suggestion
* If you select *Cron*, enter the cron expression in the *Expression* field and the time zone in the *Time Zone* field. 
For
```

>  the format and valid values for the *Time Zone* field, see the Javadoc for the Java TimeZone class.

Reviewer 5: Where is the Javadoc?

Reviewer 4: Why does the first choice get bullets but the second choice get narrative treatment? They each have more than one field to supply a value for. And do we have to define the field more granularly than its label implies?

Author: I don't know what you mean.

Reviewer 4: First: 
"If fixed...:
 * Frequency
 * Start
 * Time
"If cron...*Expression* and...*Timezone*

Reviewer 4: And do we really have to define Frequency, Start, Time, Expression, and Timezone, or are their labels self-evident?

+
. To configure metadata for the payload, variables, and attributes, select the *Metadata* tab and add the appropriate metadata.

> For more information about Scheduler, see xref:mule-runtime::scheduler-concept.adoc[Scheduler Endpoint (Trigger)].

Reviewer 4: ditto previous question about "trigger"
Also, I saw this in a title above, and did not comment because I did not think the cross-ref was to a connector topic, but titles should avoid using any punctuation, including parentheses. The modification should come in the text following the title rather than in the title itself, at least if it requires punctuation.

Author: This is the title of the topic I am referring to.

Author: I don't know what you mean by the first choice getting bullets and the second choice getting narrative treatment.

Reviewer 4: See explanation above.

> === Use a Connector-Specific Input Source

Reviewer 5: I don't think this should be in an intro topic. I think the only thing that should be in the intro is conceptual information. Procedures should be with the connector docs.

Author: Again, I am trying to document a general path for configuring a connector. When the user guides are beefed up, we can consider removing this.

If a connector has associated input sources, you can use a connector-specific input source to initiate the flow. Each connector input source has a unique configuration.

> For example, you can use Amazon SQS Receive Messages as an input source for Amazon SQS Connector. This input source initiates a flow each time the specified Amazon SQS queue receives a new message.

Reviewer 4: ```suggestion
For example, you can use Amazon SQS Receive Messages as an input source for Anypoint Connector for Amazon SQS. This input source initiates a flow each time the specified Amazon SQS queue receives a new message.
```

To configure a connector-specific input source:

> . In the Mule Palette view, search for the connector and select the input source.

Reviewer 4: ```suggestion
. In the *Mule Palette* view, search for the connector and select the input source.
```

+
For example, search for *Amazon SQS* and select the *Receive Messages* input source:
+
.Searching for the Amazon SQS Receive Messages input source
image::introduction/intro-config-studio-select-specific.png[Amazon SQS Receive Messages input source]
+
. Drag the input source onto the canvas.
. In the configuration screen, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for the input source.
. Enter values for the required fields.

> . On the *Advanced* tab next to the *Proxy* tab, specify a reconnection strategy.

Reviewer 4: If the audience can readily find the Proxy tab, why could it not find the Advanced tab? Not sure how naming something next to another helps to locate it....

. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.

[[configureConnector]]
== Configure the Connector

To configure a connector in your Mule app, first configure a global element that can be used by all instances of that connector in the flow:

> . In the Mule Palette view, search for the connector and select the desired operation.

Reviewer 4: ```suggestion
. In the *Mule Palette* view, search for the connector and select the desired operation.
```

> . Drag the operation onto the canvas to the right of the input source. In the configuration screen for the operation, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for connector.

Reviewer 4: s/b two steps
```suggestion
. Drag the operation onto the canvas to the right of the input source. 
. In the configuration screen for the operation, click the plus sign (*+*) next to the *Connector configuration* field to add a global element for the connector.
```

. Specify the global element properties for the connector.
. On the *Advanced* tab in the global element properties, specify a reconnection strategy, if supported.
. Click *Test Connection* to confirm that Mule can connect with the specified server.
. Click *OK*.

After you configure a global element, configure the other fields for the connector.

For information about configuring a specific connector, see the user guide and reference guide for that connector.

== See Also

> 

Reviewer 4: This topic repeats some of the information in the Overview topic. You might take a look at the overview and be sure that it does not contain specifics that belong better in the specific configuration task topics.

Author: I will.
